topdown/cache/cache.go

```diff
@@ -165,3 +171,14 @@ func (c *cache) maxSizeBytes() int64 {
 	}
 	return *c.config.InterQueryBuiltinCache.MaxSizeBytes
 }
+
+type keyMutex struct {
+	sync.Map
```

Comment:
It's not that I don't trust your judgement, it's just that if this design was informed by some specific piece of documentation, or anything, I'd appreciate if we recorded that 🤔 (To put it in another way: I'm not sure I would trust myself coming up with this 🙃)

Reply:
I guess what I'm wondering about is that this feels like we're locking twice. We're storing Mutexes in a Mutex-protected Map.

> The Map type is optimized for two common use cases: 
>   (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or
>   (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

So, wondering which of these would be the case here....  once we've added a Mutex for a specific `k`, it'll never be overwritten. So I guess we're in (1) territory.

But I'm not entirely sure, since we're also caching the results of http.send, and this is just the sync map guarding their re-fetches... we're basically forcing each http.send call with the same arguments (`k`) into a line: only of one them can ever be evaluated at the same time. 🤔

Reply:
The idea was to lock per key access and I started with a map + lock/unlock and then discovered `sync.Map` which seemed to work. I still trying to figure out how to test this but there isn't much in terms of documentation to point to.

Reply:
> we're basically forcing each http.send call with the same arguments (k) into a line: only of one them can ever be evaluated at the same time.

This should affect concurrent requests for the same key. Are you thinking of any side-effect this may have.

Reply:
Isn't that what were doing? I can only imagine this making things better, since the second one in line will get an immediate response when the first one is done. And every subsequent one, too, until the cage entry is stale... But I might be missing something.

Reply:
> Isn't that what were doing? 

We were protecting the cache implementation itself from concurrent access. But there could be concurrent calls to http.send wherein it's possible the first writes an entry to the cache for key `x` and the second does not see that and adds an entry for `x` too in the cache instead of just fetching it.